web/pdf_find_controller.js

```diff
@@ -331,193 +462,140 @@ class PDFFindController {
+        match.index,
+        match[0].length
+      );
+      matches.push(matchPos);
+      matchesLength.push(matchLen);
     }
     this._pageMatches[pageIndex] = matches;
     this._pageMatchesLength[pageIndex] = matchesLength;
   }
 
-  _calculateWordMatch(query, pageIndex, pageContent, pageDiffs, entireWord) {
-    const matchesWithLength = [];
-
-    // Divide the query into pieces and search for text in each piece.
-    const queryArray = query.match(/\S+/g);
-    for (let i = 0, len = queryArray.length; i < len; i++) {
-      const subquery = queryArray[i];
-      const subqueryLen = subquery.length;
-
-      let matchIdx = -subqueryLen;
-      while (true) {
-        matchIdx = pageContent.indexOf(subquery, matchIdx + subqueryLen);
-        if (matchIdx === -1) {
-          break;
+  _convertToRegExpString(query, hasDiacritics) {
+    const { matchDiacritics } = this._state;
+    query = query.replace(
+      SPECIAL_CHARS_REG_EXP,
+      (
+        match,
+        p1 /* to escape */,
+        p2 /* punctuation */,
+        p3 /* whitespaces */,
+        p4 /* diacritics */,
+        p5 /* letters */
+      ) => {
+        if (p1) {
+          // Escape characters like *+?... to not interfer with regexp syntax.
+          return `\\s*\\${p1}\\s*`;
         }
-        if (
-          entireWord &&
-          !this._isEntireWord(pageContent, matchIdx, subqueryLen)
-        ) {
-          continue;
+        if (p2) {
+          // Allow whitespaces around punctuation signs.
+          return `\\s*${p2}\\s*`;
+        }
+        if (p3) {
+          // Replace spaces by \s+ to be sure to match any spaces.
+          return "\\s+";
+        }
+        if (matchDiacritics) {
+          return p4 || p5;
         }
-        const originalMatchIdx = getOriginalIndex(matchIdx, pageDiffs),
-          matchEnd = matchIdx + subqueryLen - 1,
-          originalQueryLen =
-            getOriginalIndex(matchEnd, pageDiffs) - originalMatchIdx + 1;
-
-        // Other searches do not, so we store the length.
-        matchesWithLength.push({
-          match: originalMatchIdx,
-          matchLength: originalQueryLen,
-          skipped: false,
-        });
-      }
-    }
 
-    // Prepare arrays for storing the matches.
-    this._pageMatchesLength[pageIndex] = [];
-    this._pageMatches[pageIndex] = [];
+        if (p4) {
+          // Diacritics are removed with few exceptions.
+          return DIACRITICS_EXCEPTION.has(p4.charCodeAt(0)) ? p4 : "";
+        }
 
-    // Sort `matchesWithLength`, remove intersecting terms and put the result
-    // into the two arrays.
-    this._prepareMatches(
-      matchesWithLength,
-      this._pageMatches[pageIndex],
-      this._pageMatchesLength[pageIndex]
+        // A letter has been matched and it can be followed by any diacritics
+        // in normalized text.
+        return hasDiacritics ? `${p5}\\p{M}*` : p5;
+      }
     );
+
+    if (matchDiacritics) {
+      // aX musn't match aXY.
```

Comment:
Spelling-nit: `musn't` -> `mustn't`, or perhaps just `must not` which is a tiny bit easier to read.